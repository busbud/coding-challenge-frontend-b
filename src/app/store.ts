import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit'

import departureSearchReducer from '../features/departure-search/departureSearchSlice'

export const store = configureStore({
  reducer: {
    departureSearch: departureSearchReducer,
  },
})

export type AppDispatch = typeof store.dispatch
export type RootState = ReturnType<typeof store.getState>
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>
