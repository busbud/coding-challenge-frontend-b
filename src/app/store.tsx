/**
 * With a similar approach than api/types, putting everything in just one file as the challenge is
 * simple (And I need easy exploration of files for clarity), but SHOULD be splitted in different directories.
 * Probably using Redux/Sagas instead of Context. Due simplicity, I opted for this.
 */

import React from 'react';
import { ResponseSearch, Departure, SearchList, ResponseCity } from 'types';
  
 interface SimpleStore {
  searches: SearchList;
  departures: Array<Departure>;
  cities: Record<string, ResponseCity>;
  lang: string;
  isDirty: boolean;
  isLoading: boolean;
  clean: () => void;
  toggleLoading: (value: boolean) => void;
  setIsDirty: (value: boolean) => void;
  setLang: (lang: string) => void;
  addSearch: (response: ResponseSearch) => void;
  setCities: (cities: Record<string, ResponseCity>) => void;
  setDepartures: (departures: Array<Departure>) => void;
}

const defaultState: SimpleStore = {
  searches: {} as SearchList,
  departures: [],
  cities: {},
  lang: "en",
  isDirty: false,
  isLoading: false,
  clean: () => undefined,
  setLang: () => undefined,
  toggleLoading: () => undefined,
  setIsDirty: () => undefined,
  addSearch: () => undefined,
  setCities: () => undefined,
  setDepartures: () => undefined,
};

const Store = React.createContext<SimpleStore>(defaultState);

export default Store;