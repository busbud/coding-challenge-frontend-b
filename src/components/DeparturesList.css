.departures-list {
  list-style-type: none;
  padding: 0;
}
.departures-list-item {
  margin-bottom: 2rem;
  font-weight: 300;
  animation: pop-in 0.5s;
}
.departures-list-item:nth-child(even) {
  animation: pop-in-inverse 0.5s;
}
.departures-list-item__row {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}
.departures-list-item__departure-time {
  color: #199ceb;
}
.departures-list-item__arrival-time {
  color: rgba(10, 59, 95, 0.54);
}

@keyframes pop-in {
  0% {
    opacity: 0;
    transform: translateX(-500px);
  }
  100% {
    opacity: 1;
    transform: translateX(0px);
  }
}

@keyframes pop-in-inverse {
  0% {
    opacity: 0;
    transform: translateX(500px);
  }
  100% {
    opacity: 1;
    transform: translateX(0px);
  }
}
