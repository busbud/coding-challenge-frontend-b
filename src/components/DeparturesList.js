import React from 'react';
import PropTypes from 'prop-types';
import { map } from 'ramda';
import Departure from './Departure';

const DeparturesList = ({ departures }) => {
  return (
    <div className="App-body">
      {map(departureInfo => <Departure key={departureInfo.id} {...departureInfo} />, departures)}
    </div>
  );
};

DeparturesList.propTypes = {
  departures: PropTypes.arrayOf(PropTypes.object),
};

DeparturesList.defaultProps = {
  departures: [],
};

export default DeparturesList;
