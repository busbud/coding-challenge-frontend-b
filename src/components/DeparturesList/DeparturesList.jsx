import React, { useContext, useRef } from "react";
import { useTranslation } from "react-i18next";

import { AppContext } from "context";
import { currencyFormatter } from "utils/price";
import { formattedDate } from "helpers/date";
import Container from "commons/Container";
import FakeLoader from "commons/FakeLoader";

import {
  DeparturesWrapper,
  OperatorImage,
  Box,
  CityBox,
  CityBoxContent,
  CityBoxTitle,
  CityBoxName,
  CityBoxTime,
  CityBoxOverlay,
  OperatorBox,
  OperatorPrice,
} from "./styles";

const DeparturesList = () => {
  const { t } = useTranslation();
  const myRef = useRef(null);
  const {
    departures = [],
    locations = [],
    operators = [],
    cities = [],
    loading = false,
    searched = false,
  } = useContext(AppContext);

  if (!searched) {
    return <></>;
  }

  if (searched && loading) {
    return (
      <DeparturesWrapper>
        <Container>
          <h2>{t("departuresSearchMessage")}</h2>
          <Box>
            <CityBox loading="true">
              <FakeLoader style={{ width: "50%", marginTop: 20 }} />
              <FakeLoader style={{ width: "90%", marginBottom: 30 }} />
              <FakeLoader style={{ width: "70%" }} />
              <FakeLoader style={{ width: "40%" }} />
            </CityBox>
            <CityBox loading="true">
              <FakeLoader style={{ width: "50%", marginTop: 20 }} />
              <FakeLoader style={{ width: "90%", marginBottom: 30 }} />
              <FakeLoader style={{ width: "70%" }} />
              <FakeLoader style={{ width: "40%" }} />
            </CityBox>
            <OperatorBox loading="true">
              <FakeLoader style={{ width: "50%", marginTop: 20 }} />
              <FakeLoader style={{ width: "90%", marginBottom: 30 }} />
              <FakeLoader style={{ width: "70%" }} />
              <FakeLoader style={{ width: "40%" }} />
            </OperatorBox>
          </Box>
        </Container>
      </DeparturesWrapper>
    );
  }

  if (
    searched &&
    !loading &&
    !departures &&
    !locations &&
    !operators &&
    !cities
  ) {
    return (
      <DeparturesWrapper>
        <Container>
          <h2>{t("departuresNotFoundMessage")}</h2>
        </Container>
      </DeparturesWrapper>
    );
  }

  if (
    !departures?.length ||
    !locations?.length ||
    !operators?.length ||
    !cities?.length
  ) {
    return (
      <DeparturesWrapper>
        <Container>
          <h2>{t("departuresErrorMessage")}</h2>
        </Container>
      </DeparturesWrapper>
    );
  }

  const reducedDepartures = departures.reduce((departuresArray, departure) => {
    const destination = locations.find(
      ({ id }) => departure.destination_location_id === id
    );
    const origin = locations.find(
      ({ id }) => departure.origin_location_id === id
    );
    const operator = operators.find(({ id }) => departure.operator_id === id);

    origin.city = cities.find(({ id }) => origin.city_id === id);
    destination.city = cities.find(({ id }) => destination.city_id === id);

    departuresArray.push({
      ...departure,
      origin,
      destination,
      operator,
    });

    return departuresArray;
  }, []);

  myRef?.current?.scrollIntoView();

  return (
    <DeparturesWrapper ref={myRef}>
      <Container>
        <h2>{t("departuresFound")}</h2>
        {reducedDepartures.map(
          ({
            id,
            operator,
            origin,
            departure_time,
            destination,
            arrival_time,
            prices,
          }) => (
            <Box key={id}>
              <CityBox image_url={origin.city.hero_image_url}>
                <CityBoxOverlay />
                <CityBoxContent>
                  <CityBoxTitle>{t("origin")}</CityBoxTitle>
                  <div>
                    <CityBoxTime>{formattedDate(departure_time)}</CityBoxTime>
                    <CityBoxName>
                      {origin.name} - {origin.city.name}
                    </CityBoxName>
                  </div>
                </CityBoxContent>
              </CityBox>
              <CityBox image_url={destination.city.hero_image_url}>
                <CityBoxOverlay />
                <CityBoxContent>
                  <CityBoxTitle>{t("destination")}</CityBoxTitle>
                  <div>
                    <CityBoxTime>{formattedDate(arrival_time)}</CityBoxTime>
                    <CityBoxName>
                      {destination.name} - {destination.city.name}
                    </CityBoxName>
                  </div>
                </CityBoxContent>
              </CityBox>
              <OperatorBox>
                <OperatorImage src={operator.logo_url} alt={operator.name} />
                <OperatorPrice>{currencyFormatter(prices.total)}</OperatorPrice>
              </OperatorBox>
            </Box>
          )
        )}
      </Container>
    </DeparturesWrapper>
  );
};

export default DeparturesList;
