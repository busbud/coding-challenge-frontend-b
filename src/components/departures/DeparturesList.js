import React from 'react'
import _ from 'underscore'

class DeparturesList extends React.Component {
  constructor(props){
    super(props)

    this.state = props
  }

  componentWillReceiveProps(nextProps){
    console.log('nextProps')
    console.log(nextProps)

    if(! (_.isEqual(this.state, nextProps))){
      this.setState({ departures: nextProps.departures })
    }
  }

  render(){
    const { departures } = this.state
    console.log(departures)

    return(
      <div>
        {
          _.map(departures, function(value){
            return(
              <h1>{value.arrival_timezone}</h1>
            )
          })
        }
      </div>
    )
  }
}

export default DeparturesList
