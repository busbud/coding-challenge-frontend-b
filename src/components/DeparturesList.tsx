import React from "react";
import { StoreState } from "../types/StoreState";
import { connect } from "react-redux";
import "./DeparturesList.css";
import IDeparture from "../types/departure";
import IOperator from "../types/operator";
import ILocation from "../types/location";
import ICity from "../types/city";
import { DeparturesListItem } from "./DeparturesListItem";
import Button from "./ui/Button";
import { sortByTime, sortByPrice } from "../actions";

interface DeparturesListProps {
  departures: Array<IDeparture>;
  locations: Array<ILocation>;
  operators: Array<IOperator>;
  cities: Array<ICity>;
  searching: boolean;
  selectText: string;
  error: string;
  sortByPrice: () => void;
  sortByTime: () => void;
  priceSortText: string;
  timeSortText: string;
}

class DeparturesList extends React.Component<DeparturesListProps, any> {
  render() {
    let listElement;
    let sortControls;
    if (this.props.error.length > 0) {
      console.log(this.props.error);
      return <div>{this.props.error}</div>;
    }
    if (this.props.searching) {
      listElement = <div>Searching...</div>;
    } else {
      // Display the controls for sorting only if there are enough items in the list to sort
      if (this.props.departures.length > 1) {
        sortControls = (
          <div className="sorting-controls">
            <Button
              type="secondary"
              size="medium"
              onClick={this.props.sortByTime}
            >
              {this.props.timeSortText}
            </Button>
            <Button
              type="secondary"
              size="medium"
              onClick={this.props.sortByPrice}
            >
              {this.props.priceSortText}
            </Button>
          </div>
        );
      }
      listElement = this.props.departures.map((departure, index) => {
        // get origin by filtering from locations
        const origin = this.props.locations.filter(
          location => location.id === departure.origin_location_id
        )[0];
        // get destination by filtering from locations
        const destination = this.props.locations.filter(
          location => location.id === departure.destination_location_id
        )[0];
        // get city name of origin
        const originCity = this.props.cities.filter(
          c => c.id === origin.city_id
        )[0];
        // get city name of destination
        const destinationCity = this.props.cities.filter(
          c => c.id === destination.city_id
        )[0];
        // get operator by filtering from operators
        const operator = this.props.operators.filter(
          op => op.id === departure.operator_id
        )[0];
        // get duration by subtracting departure from arrival
        const duration =
          Math.abs(
            new Date(departure.arrival_time).getTime() -
              new Date(departure.departure_time).getTime()
          ) / 1000;
        return (
          <DeparturesListItem
            departure={departure}
            operator={operator}
            origin={origin}
            originCity={originCity}
            destination={destination}
            destinationCity={destinationCity}
            duration={duration}
            selectText={this.props.selectText}
            index={index}
            key={index}
          />
        );
      });
    }

    return (
      <div>
        {sortControls}
        <ul className="departures-list">{listElement}</ul>
      </div>
    );
  }
}

export default connect(
  ({
    searching,
    departures,
    locations,
    operators,
    cities,
    error
  }: StoreState) => {
    return {
      searching,
      departures,
      locations,
      operators,
      cities,
      error
    };
  },
  (dispatch: any) => {
    return {
      sortByPrice: () => dispatch(sortByPrice()),
      sortByTime: () => dispatch(sortByTime())
    };
  }
)(DeparturesList);
