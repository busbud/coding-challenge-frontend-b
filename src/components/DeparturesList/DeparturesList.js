import React, { PropTypes } from 'react';
import DepartureContainer from '../Departure/DepartureContainer';
import Loader from '../Loader/Loader';
import { orderBy }from 'lodash';

import './DeparturesList.scss';

class DeparturesList extends React.Component {
  constructor(props) {
    super(props);
  }
  getSortedDepartures(departures, sort) {
    //return an empty array if there are no departures
    if (!departures || !departures.length) {
      return [];
    }

    try {
      //sort depending on the sort type
      switch (sort) {
        case 'departureDate': 
          //order by dateDeparture ASC
          return orderBy(departures, 'display.departure_time');
        case 'price': 
          //order by price ASC
          return orderBy(departures, 'display.price');
        case 'company': 
          //order by operator's name ASC
          return orderBy(departures, 'display.operator.display_name');
        default:
        return departures;
      }
    } catch(err) {
      //in case of error (if we were not able to get all necessary parameters, just return the original array)
      return departures;
    }

  }
  render() {
    const { isFetching, data, sort } = this.props;

    //get sorted departures
    let departures = this.getSortedDepartures(data.departures, sort);

    return (
      <div className="departures-list">
        {departures.map((departure) => {
            return <DepartureContainer key={departure.id} departure={departure.display} />;
        })}
        {isFetching ? <Loader /> : null }
      </div>
    );
  }
}

DeparturesList.propTypes = {
    isFetching: PropTypes.bool.isRequired,
    data: PropTypes.object.isRequired,
    sort: PropTypes.string.isRequired
};

export default DeparturesList;