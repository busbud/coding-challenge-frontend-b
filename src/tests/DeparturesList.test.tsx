import React from "react";
import ReactDOM from "react-dom";
import DeparturesList from "../components/DeparturesList";
import { Provider } from "react-redux";
import store from "../store";
import { mockDeparture, mockLocation, mockOperator, mockCity } from "./mocks";
import renderer from "react-test-renderer";

import configureStore from "redux-mock-store";

const mockStore = configureStore();

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <Provider store={store}>
      <DeparturesList
        selectText="Search"
        priceSortText="Sort by price"
        timeSortText="Sort by time"
      />
    </Provider>,
    div
  );
});

it("renders a list when loadDepartures() is fired", () => {
  const store = mockStore({
    language: "en",
    searching: false,
    departures: [mockDeparture],
    locations: [mockLocation],
    operators: [mockOperator],
    cities: [mockCity],
    error: ""
  });
  const result = renderer
    .create(
      <Provider store={store}>
        <DeparturesList
          selectText="Search"
          priceSortText="Sort by price"
          timeSortText="Sort by time"
        />
      </Provider>
    )
    .toJSON();
  expect(result).toMatchSnapshot();
});

it("renders 'Searching...' when searching=true", () => {
  const store = mockStore({
    language: "en",
    searching: true,
    departures: [],
    locations: [],
    operators: [],
    cities: [],
    error: ""
  });
  const result = renderer
    .create(
      <Provider store={store}>
        <DeparturesList
          selectText="Search"
          priceSortText="Sort by price"
          timeSortText="Sort by time"
        />
      </Provider>
    )
    .toJSON();
  expect(result).toMatchSnapshot();
});

it("renders 'Searching...' when searching=true, despite departures being loaded", () => {
  const store = mockStore({
    language: "en",
    searching: true,
    departures: [mockDeparture],
    locations: [mockLocation],
    operators: [mockOperator],
    cities: [mockCity],
    error: ""
  });
  const result = renderer
    .create(
      <Provider store={store}>
        <DeparturesList
          selectText="Search"
          priceSortText="Sort by price"
          timeSortText="Sort by time"
        />
      </Provider>
    )
    .toJSON();
  expect(result).toMatchSnapshot();
});

it("renders error message when there's an error", () => {
  const store = mockStore({
    language: "en",
    searching: false,
    departures: [mockDeparture],
    locations: [mockLocation],
    operators: [mockOperator],
    cities: [mockCity],
    error: "Test error message"
  });
  const result = renderer
    .create(
      <Provider store={store}>
        <DeparturesList
          selectText="Search"
          priceSortText="Sort by price"
          timeSortText="Sort by time"
        />
      </Provider>
    )
    .toJSON();
  expect(result).toMatchSnapshot();
});

it("Does not try to render list if departures are loaded, but other data such as locations isn't", () => {
  const store = mockStore({
    language: "en",
    searching: false,
    departures: [mockDeparture],
    locations: [],
    operators: [],
    cities: [],
    error: ""
  });
  const result = renderer
    .create(
      <Provider store={store}>
        <DeparturesList
          selectText="Search"
          priceSortText="Sort by price"
          timeSortText="Sort by time"
        />
      </Provider>
    )
    .toJSON();
  expect(result).toMatchSnapshot();
});
