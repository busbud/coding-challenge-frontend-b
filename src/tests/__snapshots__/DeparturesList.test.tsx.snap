// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Does not try to render list if departures are loaded, but other data such as locations isn't 1`] = `
<div>
  <ul
    className="departures-list"
  />
</div>
`;

exports[`renders 'Searching...' when searching=true 1`] = `
<div>
  <ul
    className="departures-list"
  >
    <div>
      Searching...
    </div>
  </ul>
</div>
`;

exports[`renders 'Searching...' when searching=true, despite departures being loaded 1`] = `
<div>
  <ul
    className="departures-list"
  >
    <div>
      Searching...
    </div>
  </ul>
</div>
`;

exports[`renders a list when loadDepartures() is fired 1`] = `
<div>
  <div
    className="sorting-controls"
  >
    <div
      className="undefined button button--secondary button--medium"
      onClick={[Function]}
    >
      Sort by time
    </div>
    <div
      className="undefined button button--secondary button--medium"
      onClick={[Function]}
    >
      Sort by price
    </div>
  </div>
  <ul
    className="departures-list"
  >
    <li
      className="departures-list-item"
      id="dli-0-Métro Bonaventure Bus Station-Métro Bonaventure Bus Station-2016-01-14T07:55:00"
    >
      <div
        className="card undefined"
      >
        <div
          className="departures-list-item__row"
        >
          <img
            alt="Greyhound"
            className="departures-list-item__img"
            src="https://busbud-pubweb-assets-staging.global.ssl.fastly.net/images-service/operator-logos/greyhound.png?hash=1{&height,width}"
          />
          <div>
            $
            52.00
          </div>
        </div>
        <div
          className="departures-list-item__row"
        >
          <div>
            <span
              className="departures-list-item__departure-time"
            >
              <b>
                00:01:00
              </b>
            </span>
             
            <span>
              New York
               - 
              Métro Bonaventure Bus Station
            </span>
          </div>
        </div>
        <div
          className="departures-list-item__row"
        >
          <div
            className="departures-list-item__arrival-time"
          >
            <span>
              <b>
                07:55:00
              </b>
            </span>
             
            <span>
              New York
               - 
              Métro Bonaventure Bus Station
            </span>
          </div>
        </div>
        <div
          className="departures-list-item__row"
        >
          <div>
            7
            h 
            54
            m
          </div>
          <div
            className="undefined button button--primary button--medium"
          >
            Search
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;

exports[`renders error message when there's an error 1`] = `
<div>
  Test error message
</div>
`;
